examples/fluid.py

```diff
@@ -0,0 +1,233 @@
+                    pos[0])
+        v_y2 = lerp(cx1, cx2, sample_v(vf, cx1, cy2), sample_v(vf, cx2, cy2),
+                    pos[0])
+        new_v = lerp(cy1, cy2, v_y1, v_y2, pos[1])
+        new_vf[i, j] = new_v
+
+
+v_alpha = dx * dx / (viscority * dt)
+v_rbeta = 1.0 / (4.0 + v_alpha)
+
+
+@ti.kernel
+def diffusion_jacobi(vf: ti.template(), new_vf: ti.template()):
+    for i, j in vf:
+        vl = sample_v(vf, i - 1, j)
+        vr = sample_v(vf, i + 1, j)
+        vt = sample_v(vf, i, j + 1)
+        vb = sample_v(vf, i, j - 1)
+
+        b = vf[i, j]
+        new_vf[i, j] = (vl + vr + vt + vb + b * v_alpha) * v_rbeta
+
+
+rforce_radius = (3.0 / res)
+f_strength_dt = f_strength * dt
+
+
+@ti.kernel
+def apply_mouse_impulse(new_vf: ti.template(), mouse_arr: ti.ext_arr()):
+    for i, j in new_vf:
+        omx, omy = mouse_arr[2], mouse_arr[3]
+        mdir = ti.Vector([mouse_arr[0], mouse_arr[1]])
+        dx, dy = (i + 0.5 - omx), (j + 0.5 - omy)
+        d2 = dx * dx + dy * dy
+        momentum = mdir * f_strength_dt * ti.exp(-d2 * rforce_radius)
+        v = new_vf[i, j]
+        v += momentum
+        new_vf[i, j] = v
+
+
+@ti.kernel
+def divergence(vf: ti.template()):
+    for i, j in vf:
+        vl = sample_v(vf, i - 1, j)[0]
+        vr = sample_v(vf, i + 1, j)[0]
+        vt = sample_v(vf, i, j + 1)[1]
+        vb = sample_v(vf, i, j - 1)[1]
+        div_vels[i, j] = ((vr - vl) + (vt - vb)) * half_rdx
+
+
+p_alpha = -dx * dx
+
+
+@ti.kernel
+def pressure_jacobi(pf: ti.template(), new_pf: ti.template()):
+    for i, j in pf:
+        pl = sample_p(pf, i - 1, j)
+        pr = sample_p(pf, i + 1, j)
+        pt = sample_p(pf, i, j + 1)
+        pb = sample_p(pf, i, j - 1)
+        div = div_vels[i, j]
+        new_pf[i, j] = (pl + pr + pt + pb + p_alpha * div) * 0.25
+
+
+@ti.kernel
+def subtract_gradient(new_vf: ti.template(), pf: ti.template()):
+    # Make |new_vf| divergence free
+    for i, j in new_vf:
+        pl = sample_p(pf, i - 1, j)
+        pr = sample_p(pf, i + 1, j)
+        pt = sample_p(pf, i, j + 1)
+        pb = sample_p(pf, i, j - 1)
+        v = new_vf[i, j]
+        v -= half_rdx * ti.Vector([pr - pl, pt - pb])
+        v *= v_decay
+        new_vf[i, j] = v
+
+
+@ti.kernel
+def fill_color(new_vf: ti.template()):
+    for i, j in new_vf:
+        v = new_vf[i, j]
+        color_buffer[i, j] = ti.Vector([ti.abs(v[0]), ti.abs(v[1]), 0.25])
+
+
+def reset():
+    velocities.fill(ti.Vector([0.0, 0.0]))
+    pressures.fill(0.0)
+    color_buffer.fill(ti.Vector([0.0, 0.0, 0.0]))
+
+
+def step(mouse_data):
+    global velocities, new_velocities, pressures, new_pressures
+    advect(velocities, new_velocities)
+    if enable_diffusion:
+        for _ in range(v_jacobi_iters):
+            diffusion_jacobi(velocities, new_velocities)
+            velocities, new_velocities = new_velocities, velocities
+    apply_mouse_impulse(new_velocities, mouse_data)
+    divergence(new_velocities)
+    pressures.fill(0.0)
```

Comment:
Actually, a common trick in Jaocbi pressure projection is *warm starting*, meaning using the pressure solution of the last frame as the initial guess. So we don't need `pressure.fill(0)` here.

Reply:
I see. Yeah that makes much more sense. I think this also made a real difference -- Now when i run this, the velocity fields propagates much faster, with divergence at the magnitude of `1e-5` *before it hits the boundary*. So i believe boundary is totally wrong..

Reply:
Right, you can play with warm starting here: http://yuanming-hu.github.io/fluid/ There's an option.